Support std c++17 builds with Windows

Currently when building the demos against Ice using C++17 I get some errors.
```
[build] c:\Users\joe\Developer\zeroc-ice\ice\cpp\include\Ice\VersionFunctions.h(20,44): error C7555: use of designated initializers requires at least '/std:c++20' [C:\Users\joe\Developer\zeroc-ice\ice-demos\build\Ice\greeterAsync\ice_greeter_async_client.vcxproj]
[build]   (compiling source file '../../../cpp/Ice/greeterAsync/Client.cpp')
[build]   
[build] C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.42.34433\include\span(12): warning STL4038: The contents of <span> are available only with C++20 or later. [C:\Users\joe\Developer\zeroc-ice\ice-demos\build\Ice\greeterAsync\ice_greeter_async_client.vcxproj]
[build] c:\Users\joe\Developer\zeroc-ice\ice\cpp\include\Ice\VersionFunctions.h(14,44): error C7555: use of designated initializers requires at least '/std:c++20' [C:\Users\joe\Developer\zeroc-ice\ice-demos\build\Ice\greeter\ice_greeter_server.vcxproj]
[build]   (compiling source file '../../../cpp/Ice/greeter/Server.cpp')
```

--- a/.clang-tidy
+++ b/.clang-tidy
@@ -7,6 +7,7 @@ Checks:
   -modernize-avoid-c-arrays,
   -modernize-use-trailing-return-type,
   -modernize-macro-to-enum,
+  -modernize-use-designated-initializers,
   performance-*,
   -performance-avoid-endl,
   -performance-enum-size,

--- a/cpp/include/DataStorm/InternalT.h
+++ b/cpp/include/DataStorm/InternalT.h
@@ -154,10 +154,7 @@ namespace DataStormI
     public:
         AbstractFactoryT() = default;
 
-        void init()
-        {
-            _deleter = Deleter{std::enable_shared_from_this<AbstractFactoryT<K, V>>::shared_from_this()};
-        }
+        void init() { _deleter = Deleter{std::enable_shared_from_this<AbstractFactoryT<K, V>>::shared_from_this()}; }
 
         template<typename F, typename... Args>
         [[nodiscard]] std::shared_ptr<typename V::BaseClassType> create(F&& value, Args&&... args)

--- a/cpp/include/DataStorm/InternalT.h
+++ b/cpp/include/DataStorm/InternalT.h
@@ -156,7 +156,7 @@ namespace DataStormI
 
         void init()
         {
-            _deleter = Deleter{._factory = std::enable_shared_from_this<AbstractFactoryT<K, V>>::shared_from_this()};
+            _deleter = Deleter{std::enable_shared_from_this<AbstractFactoryT<K, V>>::shared_from_this()};
         }
 
         template<typename F, typename... Args>

--- a/cpp/include/Ice/StreamHelpers.h
+++ b/cpp/include/Ice/StreamHelpers.h
@@ -12,7 +12,7 @@
 
 #if defined(_MSC_VER)
 // With MSVC, __has_include(<span>) evaluates to 1 in c++17 mode and the <span> header then issues a warning. This
-// is a permissible but impractical behavior. As a result, we include <span> only in c++20 mode.
+// is a permissible but impractical behavior. As a result, we include <span> only for c++20 or higher.
 #    if _MSVC_LANG >= 202002L
 #        include <span>
 #    endif

--- a/cpp/include/Ice/StreamHelpers.h
+++ b/cpp/include/Ice/StreamHelpers.h
@@ -10,7 +10,11 @@
 #include <iterator>
 #include <ostream>
 
-#if __has_include(<span>)
+#if defined(_MSC_VER) // workaround for MSVC bug
+#    if _MSVC_LANG >= 202002L
+#        include <span>
+#    endif
+#elif __has_include(<span>)
 #    include <span>
 #endif
 

--- a/cpp/include/Ice/StreamHelpers.h
+++ b/cpp/include/Ice/StreamHelpers.h
@@ -11,8 +11,8 @@
 #include <ostream>
 
 #if defined(_MSC_VER)
-// With MSVC, __has_include(<span>) evaluates to 1 in c++17 mode and the <span> header then issues a warning. This
-// is a permissible but impractical behavior. As a result, we include <span> only for c++20 or higher.
+// With MSVC, __has_include(<span>) evaluates to 1 in c++17 mode and the <span> header then issues a warning that can't
+// be disabled. This is a permissible but impractical behavior. See https://github.com/microsoft/STL/issues/4010.
 #    if _MSVC_LANG >= 202002L
 #        include <span>
 #    endif

--- a/cpp/include/Ice/StreamHelpers.h
+++ b/cpp/include/Ice/StreamHelpers.h
@@ -10,7 +10,9 @@
 #include <iterator>
 #include <ostream>
 
-#if defined(_MSC_VER) // workaround for MSVC bug
+#if defined(_MSC_VER)
+// With MSVC, __has_include(<span>) evaluates to 1 in c++17 mode and the <span> header then issues a warning. This
+// is a permissible but impractical behavior. As a result, we include <span> only in c++20 mode.
 #    if _MSVC_LANG >= 202002L
 #        include <span>
 #    endif

--- a/cpp/include/Ice/VersionFunctions.h
+++ b/cpp/include/Ice/VersionFunctions.h
@@ -11,13 +11,13 @@
 namespace Ice
 {
     /** Identifies protocol version 1.0. */
-    constexpr ProtocolVersion Protocol_1_0{.major = 1, .minor = 0};
+    constexpr ProtocolVersion Protocol_1_0{1, 0};
 
     /** Identifies encoding version 1.0. */
-    constexpr EncodingVersion Encoding_1_0{.major = 1, .minor = 0};
+    constexpr EncodingVersion Encoding_1_0{1, 0};
 
     /** Identifies encoding version 1.1. */
-    constexpr EncodingVersion Encoding_1_1{.major = 1, .minor = 1};
+    constexpr EncodingVersion Encoding_1_1{1, 1};
 
     /** Identifies the latest protocol version. */
     constexpr ProtocolVersion currentProtocol{Protocol_1_0};

